src/fanControl.ino

```diff
@@ -0,0 +1,279 @@
+   * 2) Adjust setOscillatorFrequency() until the PWM update frequency is the
+   *    expected value (50Hz for most ESCs)
+   * Setting the value here is specific to each individual I2C PCA9685 chip and
+   * affects the calculations for the PWM update frequency. 
+   * Failure to correctly set the int.osc value will cause unexpected PWM results
+   */
+  pwm.setOscillatorFrequency(27000000);
+  pwm.setPWMFreq(1600);  // This is the maximum PWM frequency
+
+  // if you want to really speed stuff up, you can go into 'fast 400khz I2C' mode
+  // some i2c devices dont like this so much so if you're sharing the bus, watch
+  // out for this!
+  Wire.setClock(400000);
+}
+
+void loop() 
+{
+  int i = 0;
+  while (Serial.available() > 0) 
+  {
+    i = i + Serial.readBytes(inBuf, 4);
+//    i++;
+//    i = 4;
+    if (i == 4) 
+    {
+      i = 0;
+//      Serial.println("4 bytes read!");
+      if (inBuf[3] == CMD_FAN_ID)
+      {
+//        Serial.println("CMD_FAN_ID");
+        // Send Fan ID
+        outBuf[3] = CMD_FAN_ID;
+        outBuf[2] = FAN_ID;
+        outBuf[1] = 0;
+        outBuf[0] = CMD_OK;
+      }
+      else if (inBuf[3] == CMD_RELAY)
+      {
+//        Serial.println("CMD_RELAY");
+        outBuf[3] = CMD_RELAY;
+        outBuf[2] = inBuf[2]; // channel
+        outBuf[1] = inBuf[1];  // relay set/reset
+        outBuf[0] = CMD_OK;
+        if ((inBuf[1] == RELAY_SET) || (inBuf[1] == RELAY_RESET))
+        {
+          if ((inBuf[2] < 14) && (inBuf[2] > 7))   // channel no.
+          {
+            digitalWrite(inBuf[2], inBuf[1]);
+          }
+          else
+          {
+            outBuf[0] = CMD_ERROR;        
+          }
+        }
+        else
+        {
+          outBuf[0] = CMD_ERROR;
+        }
+      }
+      else if (inBuf[3] == CMD_SPEED)
+      {
+  //      Serial.println("CMD_SPEED");
+        noInterrupts(); // I think we need this to successfully copy the input
+        channel = inBuf[2];
+        duty = inBuf[1];
+        interrupts();
+        outBuf[3] = CMD_SPEED;
+        outBuf[2] = channel;
+        outBuf[1] = duty;
+        outBuf[0] = CMD_OK;
+
+        if (channel < 16) // There are 16 channels on the Adafruit PWM board
+        {
+          if (duty == 0)
+          {
+            // PWM fully off command, fully on is pwm.setPWM(pin, 4096, 0)
+            pwm.setPWM(channel, 0, 4096);
+          }
+          else
+          {
+            pwm.setPWM(channel, 0, (float) duty/254.0*4095);
+          }
+        }
+        else
+        {
+          outBuf[0] = CMD_ERROR;
+        }
+      }
+      else
+      {
+//        Serial.println("Other command");
+        outBuf[3] = 6;
+        outBuf[2] = 0;
+        outBuf[1] = 0;
+        outBuf[0] = CMD_ERROR;        
+      }  
+    }
+    else
+    {
+//      Serial.println("CMD_UNKNOWN");
+      outBuf[3] = 7;
+      outBuf[2] = 0;
+      outBuf[1] = 0;
+      outBuf[0] = CMD_ERROR;
+    }
+    Serial.write(outBuf, 4);  // Send command response 

```

Comment:
Currently, in the python app, there is no function linked to these returned values. But it would be cool to test in the app if the request errored somehow and then either retry or reset the relay or some such.

There is quit a bit of error and failure handling in this code. Considering that we control the app that sends the bytes to this controller, I'd imagine we could skip most of these checks.

Reply:
Well, I was learnt to make sure there are no possibilities to send the wrong command/responses from either side. Two modules with a waterproof interface in between. ;-)